server/topdf/topdf.go

```diff
@@ -0,0 +1,143 @@
+package topdf
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+)
+
+// toPDFPrefix used as a prefix while using KV store to access PDF files' fileID.
+const toPDFPrefix = "pdf:"
+
+// ErrUnauthorizedUser returned when user has no access to a file that requested to be converted to PDF.
+var ErrUnauthorizedUser = errors.New("user is not authorized to access pdf")
+
+// TOPDF is an application that converts files to PDFs and permanently caches them by using Mattermost APIs.
+type TOPDF struct {
+	// mapi is Mattermost's Plugin API.
+	mapi plugin.API
+
+	// server used to convert files to PDF.
+	server PDFServer
+}
+
+// New creates a new TOPDF app with mapi and PDF server.
+func New(mapi plugin.API, server PDFServer) *TOPDF {
+	return &TOPDF{
+		mapi:   mapi,
+		server: server,
+	}
+}
+
+// CheckServerStatus checks if underlying PDF server is running and ready to accept requests.
+func (t *TOPDF) CheckServerStatus() (running bool, err error) {
+	return t.server.Status()
+}
+
+// GetPDF gets PDF for fileID that belongs userID. user has to have access to the file
+// otherwise ErrUnauthorizedUser is returned.
+func (t *TOPDF) GetPDF(userID, fileID string) (pdf io.ReadCloser, err error) {
+	pdf, err = t.getPDF(userID, fileID)
+	if err != nil {
+		// return an authorization error if we got an err from Plugin's API.
+		if _, ok := err.(*model.AppError); ok {
+			return nil, ErrUnauthorizedUser
+		}
+		return nil, err
+	}
+	return pdf, nil
+}
+
+// getPDF gets PDF for fileID that belongs to userID.
+// notes:
+// - Mattermost's Plugin API does not implement io.Reader while dealing with files but this might
+//   be improved in future since large files can pump memory usage. TOPDF created streams in mind,
+//   this is why we use ioutil.NopCloser and bytes.Reader in the code below, to even with current Plugin API.
+// - Mattermost's Plugin API does not return errors as `error`s but returns them as *model.AppError,
+//   this needs to be improved since it causes issues while dealing with errors. For more info
+//   please see: https://golang.org/doc/faq#nil_error
+//   to workaround this, Plugin errors are normalized with normalizeAppErr().
+func (t *TOPDF) getPDF(userID, fileID string) (pdf io.ReadCloser, err error) {
+	// try to get id of PDF file that possibly generated and cached for fileID before.
+	pid, aerr := t.mapi.KVGet(key(fileID))
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// if there is no PDF file cached, create it, cache and use its content.
+	if len(pid) == 0 {
+		data, err := t.createAndSavePDF(fileID, userID)
+		if err != nil {
+			return nil, err
+		}
+		return ioutil.NopCloser(bytes.NewReader(data)), nil
+	}
+	// we have the PDF version in cache, directly return it back.
+	data, aerr := t.mapi.GetFile(string(pid))
```

Comment:
We should verify the user has access to the file here as well, otherwise we will serve up the PDF after someone with access generates it.

Reply:
Ops, my bad. I'll be fixing this one!